lib/streamlit/runtime/scriptrunner/script_runner.py

```diff
@@ -442,6 +443,15 @@ def _run_script(self, rerun_data: RerunData) -> None:
             query_string=rerun_data.query_string,
             page_script_hash=page_script_hash,
         )
+        url = parse.parse_qs(rerun_data.query_string, keep_blank_values=True)
+        query_params = ctx.session_state._state._query_params
+        for key, val in url.items():
+            if len(val) == 0:
+                query_params.set_with_no_forward_msg(key, val="")
+            if len(val) == 1:
+                query_params.set_with_no_forward_msg(key, val=val[-1])
+            else:
+                query_params.set_with_no_forward_msg(key, val)
```

Comment:
Technically when they do something like `st.query_params.URL_PARAM`, it would be the same. However, I think when do they do `st.write(st.query_params)`, then it would be a list and I think it's preferable to actually provide "more correct" state. For example, take these e2e tests, 

As a result, I have changed the second if statement to an `elif`. Let me know what you think.

https://github.com/streamlit/streamlit/assets/103006371/8dab2e26-57c4-466c-bcfe-63337f37d34c